{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-09T03:42:59.632688Z",
     "start_time": "2021-10-09T03:42:56.627635Z"
    }
   },
   "outputs": [],
   "source": [
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/10/13 上午10:17:24\n",
    "#@Version: v2\n",
    "#@Description: \n",
    "#base on the single read met file generated by nanopolish \n",
    "#batch process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# generete the met count on each chimeric fragments on read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-13T04:17:21.164621Z",
     "start_time": "2021-10-13T04:17:18.618118Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-13T02:51:54.706491Z",
     "start_time": "2021-10-13T02:49:46.267430Z"
    }
   },
   "outputs": [],
   "source": [
    "#the single read methylation file generated by nanopolish\n",
    "met_read = pd.read_table('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/rep2_part.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-13T05:45:41.077762Z",
     "start_time": "2021-10-13T05:44:51.477965Z"
    }
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from pathlib import PurePath\n",
    "import pyarrow\n",
    "import pyarrow.parquet as pq\n",
    "import numpy as np\n",
    "from concurrent import futures\n",
    "porec_dir = '/public/home/lizw/task/pore_c/porec_1000_filter_mainchr_result'\n",
    "prefix = 'DpnII_run06'\n",
    "align_dir = Path(PurePath(porec_dir,'align_table'))\n",
    "align_dir_porec_generator = align_dir.glob(prefix + \"*pore_c.parquet\")\n",
    "align_batch_porec_merge_df = pyarrow.concat_tables(list(map(pq.read_table,align_dir_porec_generator))).to_pandas()\n",
    "align_batch_porec_merge_df_pass = align_batch_porec_merge_df.query('pass_filter==True')\n",
    "align_batch_porec_merge_df_pass_reindex = align_batch_porec_merge_df_pass.reindex(['chrom','start','end','strand','read_name'],axis=1)\n",
    "\n",
    "met_read = pd.read_feather('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/rep2_part.feather')\n",
    "sample_reads = list(met_read['read_name'].drop_duplicates().sample(10000))\n",
    "sample_read_table = align_batch_porec_merge_df_pass_reindex.query('read_name in @sample_reads')\n",
    "sample_read_table.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_read_table.csv',index=None)\n",
    "sample_met_table = met_read.query('read_name in @sample_reads')\n",
    "sample_met_table.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_met_table.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-10-13T05:47:27.784976Z",
     "start_time": "2021-10-13T05:47:27.609455Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_fragment_met_level.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile \"/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_fragment_met_level.py\"\n",
    "import pandas as pd\n",
    "sample_read_table = pd.read_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_read_table.csv')\n",
    "sample_met_table = pd.read_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_met_table.csv')\n",
    "\n",
    "#A positive value in the log_lik_ratio column indicates support for methylation\n",
    "#all are CG, need not to be identified\n",
    "#some are single nucleus acid resolution, but some are not. The judge are for all the CG inculded in the sequence.\n",
    "#the context call across two framgment would be abandoned\n",
    "\n",
    "cpu_count = 30\n",
    "batch_size = len(sample_read_table) // cpu_count\n",
    "batch_dict = dict(tuple(sample_read_table.groupby(np.arange(len(sample_read_table))//batch_size)))\n",
    "#process or thread?\n",
    "def multi_thread(func,iter_object,max_workers):\n",
    "    workers = min(max_workers,len(iter_object))\n",
    "    with futures.ThreadPoolExecutor(workers) as excutors:\n",
    "        res = excutors.map(func,iter_object)\n",
    "    return list(res)\n",
    "\n",
    "def met_to_fragment_batch(key):\n",
    "    df = batch_dict[key]\n",
    "    met_level_list = []\n",
    "    for items in df.itertuples(index=False):\n",
    "        chrom,start,end,strand,read_name = items\n",
    "        chrom_int = int(chrom)\n",
    "        sub_df = sample_met_table.query('(read_name == @read_name) & (chromosome == @chrom_int) & (start >= @start) & (end < @end)')\n",
    "        #according to Bing Ren: Only reads containing 2 or more CpGs on each end were included\n",
    "        CG_count = sub_df['num_motifs'].sum()\n",
    "        if CG_count >= 2:\n",
    "            met_sub_df = sub_df.query('log_lik_ratio > 0')\n",
    "            met_sub_value = met_sub_df['num_motifs'].sum()/CG_count\n",
    "        else:\n",
    "            met_sub_value = np.nan\n",
    "        \n",
    "        met_level_list.append(met_sub_value)\n",
    "    \n",
    "    df['met_level'] = met_level_list\n",
    "    return df\n",
    "\n",
    "met_and_read_df = pd.concat(multi_thread(met_to_fragment_batch,list(batch_dict.keys()),cpu_count),axis=0)\n",
    "met_and_read_df.to_csv('/public/home/lizw/task/pore_c/methylation/met_structure/sample_fragment_met_levels.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# split into pair-wise contat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations\n",
    "\n",
    "met_level_read = pd.read_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/sample_fragment_met_levels.csv')\n",
    "read_table_met_value_no_na = met_level_read.dropna()\n",
    "twoD_df = pd.DataFrame([])\n",
    "for i in tqdm(read_table_met_value_no_na.groupby(by='read_name')):\n",
    "    read_name,sub_df = i\n",
    "    if len(sub_df) >= 2:\n",
    "        index_list = sub_df.index\n",
    "        index_cc_list = list(combinations(index_list,2))\n",
    "        for cc in index_cc_list:\n",
    "            index1,index2 = cc\n",
    "            df_cc1 = pd.DataFrame(sub_df.loc[index1,:]).T\n",
    "            df_cc2 = pd.DataFrame(sub_df.loc[index2,:]).T\n",
    "            df_cc1.reset_index(drop=True, inplace=True)\n",
    "            df_cc2.reset_index(drop=True, inplace=True)\n",
    "            contact_sub_df = pd.concat([df_cc1,df_cc2],axis=1,ignore_index=True)\n",
    "            twoD_df = twoD_df.append(contact_sub_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# genomic distance filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "twoD_df.columns=['drop1','chrom1','start1','end1','strand1','read_name1','met_level1','drop2','chrom2','start2','end2','strand2','read_name2','met_level2']\n",
    "twoD_df.reset_index(inplace=True)\n",
    "twoD_df.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/2D_filtered.csv')\n",
    "\n",
    "twoD_df = pd.read_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/2D_filtered.csv')\n",
    "\n",
    "twoD_df['start1'] = twoD_df['start1'].astype(int)\n",
    "twoD_df['start2'] = twoD_df['start2'].astype(int)\n",
    "twoD_df['end1'] = twoD_df['end1'].astype(int)\n",
    "twoD_df['end2'] = twoD_df['end2'].astype(int)\n",
    "\n",
    "twoD_df['pos1'] = twoD_df.eval('(start1+end1)//2',engine='python')\n",
    "twoD_df['pos2'] = twoD_df.eval('(start2+end2)//2',engine='python')\n",
    "\n",
    "#distance_filter = 1_000_000\n",
    "\n",
    "twoD_df_trans = twoD_df.query('chrom1 != chrom2')\n",
    "twoD_df_cis = twoD_df.query('chrom1 == chrom2')\n",
    "twoD_df_cis_dis = twoD_df_cis.assign(distance = lambda df:abs(df['pos2']-df['pos1']))\n",
    "# twoD_df_cis_filter_1k = twoD_df_cis_dis.query('distance >= 1000')\n",
    "# twoD_df_cis_filter_5k = twoD_df_cis_dis.query('distance >= 5000')\n",
    "twoD_df_cis_filter_25k = twoD_df_cis_dis.query('distance >= 25000')\n",
    "twoD_df_cis_filter_100k = twoD_df_cis_dis.query('distance >= 100000')\n",
    "twoD_df_cis_filter_1M = twoD_df_cis_dis.query('distance >= 1000000')\n",
    "\n",
    "twoD_df_cis_filter_25k.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/twoD_df_cis_filter_25k.csv')\n",
    "twoD_df_cis_filter_100k.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/twoD_df_cis_filter_100k.csv')\n",
    "twoD_df_cis_filter_1M.to_csv('/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/twoD_df_cis_filter_1M.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# shuffle generating"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def met_contact_corr(twoD_df_cis_filter):\n",
    "    size = len(twoD_df_cis_filter)\n",
    "    inread_met_level1 = twoD_df_cis_filter['met_level1'].astype('float')\n",
    "    inread_met_level2 = twoD_df_cis_filter['met_level2'].astype('float')\n",
    "    read_r = inread_met_level1.corr(inread_met_level2)\n",
    "    import random\n",
    "    from scipy import stats\n",
    "    shuffle_met_level1 = list(inread_met_level1)\n",
    "    shuffle_met_level2 = list(inread_met_level2)\n",
    "    random.shuffle(shuffle_met_level1)\n",
    "    random.shuffle(shuffle_met_level2)\n",
    "    shuffle_r, shuffle_p = stats.pearsonr(shuffle_met_level1, shuffle_met_level2)\n",
    "    return read_r,shuffle_r,size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_25k = dict(zip(('Long-C contact','shuffle','size'),met_contact_corr(twoD_df_cis_filter_25k)))\n",
    "dict_100k = dict(zip(('Long-C contact','shuffle','size'),met_contact_corr(twoD_df_cis_filter_100k)))\n",
    "dict_1M = dict(zip(('Long-C contact','shuffle','size'),met_contact_corr(twoD_df_cis_filter_1M)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%config InlineBackend.figure_format = 'retina'\n",
    "corr_list = [dict_25k,dict_100k,dict_1M]\n",
    "\n",
    "corr_df_all = pd.DataFrame([])\n",
    "for i in corr_list:\n",
    "    read_r = i['Long-C contact']\n",
    "    shuffle_r = i['shuffle']\n",
    "    corr_table = pd.DataFrame({'Long-C Contact':[read_r],'Shuffled':[shuffle_r]})\n",
    "    corr_df_all = corr_df_all.append(corr_table)\n",
    "\n",
    "corr_df_all.index = ['25kb','100kb','1Mb']\n",
    "import matplotlib.pyplot as plt\n",
    "from cycler import cycler\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.font_manager as font_manager\n",
    "import matplotlib as mpl\n",
    "import matplotlib.patches as mp\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "from cycler import cycler\n",
    "#for more\n",
    "mpl.rcParams['axes.prop_cycle'] = cycler('color',['#B2957A','#A0A0A0'])\n",
    "#for three\n",
    "\n",
    "# 设置全局字体\n",
    "font_dirs = ['/public/home/mowp/test/fonts/']\n",
    "font_files = font_manager.findSystemFonts(fontpaths=font_dirs)\n",
    "\n",
    "for font_file in font_files:\n",
    "    font_manager.fontManager.addfont(font_file)\n",
    "\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams['font.size'] = 15\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "fig,ax = plt.subplots(figsize=(3,4))\n",
    "corr_df_all.loc[('25kb','100kb','1Mb'),:].plot(kind = 'bar',ax=ax,width = 0.7)\n",
    "sns.despine(ax=ax)\n",
    "ax.legend(loc=2,bbox_to_anchor=(1.05,1.0),borderaxespad=0)\n",
    "plt.savefig(f'/public/home/lizw/task/pore_c/methylation/nanopolish/rep2/all_filter.png',bbox_inches='tight',dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
