{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-12-20T07:53:27.387660Z",
     "start_time": "2021-12-20T07:53:27.346231Z"
    }
   },
   "outputs": [],
   "source": [
    "#不运行参数定义的行,吧参数定义行和实体运行行分开，把args.改成args.就能运行例子而不是写py\n",
    "args.region = '/public/home/lizw/task/pore_c/porec_result2/bed/all_chr.bed'\n",
    "args.anchor = '/public/home/lizw/task/pore_c/porec_result2/bed/peri.all.bed'\n",
    "args.align_merge_part = '/public/home/lizw/task/pore_c/porec_1000_filter_mainchr_result/porecplot/DpnII_run04.order_2_merge_part.csv'\n",
    "args.keepall = '/public/home/lizw/task/pore_c/porec_1000_filter_mainchr_result/porecplot/DpnII_run04.order_2_sorted_orderall_fornormailize.csv'\n",
    "args.clusterheat = True\n",
    "args.binsize = 500000\n",
    "args.calculate = 'raw'\n",
    "args.prefix = 'test1'\n",
    "args.mark = '/public/home/lizw/task/pore_c/pipeline_test_v1/setfiles/test_mark.csv'\n",
    "args.bedgraphs = ['/public/home/lizw/data/IMPORTANT_genome/arabidopsis/modification/H3K4me2-DRX066771.bw.bg','/public/home/lizw/data/IMPORTANT_genome/arabidopsis/modification/H3K27me3-SRX648279.bw.bg']\n",
    "args.bg_labels = ['H3K4me2','H3K27me3']\n",
    "args.bg_ylims = [None,None]\n",
    "args.plotkind = 'bar'\n",
    "args.sites = []\n",
    "args.site_labels = [None]\n",
    "args.site_ylims = [None]\n",
    "args.anchormode = 'and'\n",
    "args.genemodel = ''\n",
    "args.slice_cluster = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-01-04T09:56:41.350083Z",
     "start_time": "2022-01-04T09:56:41.298651Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing /public/home/lizw/task/pore_c/tools/porecscripts/v2.15/porecplot_draw.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile '/public/home/lizw/task/pore_c/tools/porecscripts/v2.15/porecplot_draw.py'\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/12/28 下午2:16:21\n",
    "#@Version: v2.14\n",
    "#@Description: change binnor calculation method\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/10/5 下午2:42:17\n",
    "#@Version: v2.12\n",
    "#@Description: change anchorcluster color\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/10/1 下午12:16:34\n",
    "#@Version: v2.11\n",
    "#@Description: generate matrix independently with heatmap draw,modify the plot specially for website\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/9/24 下午3:37:52\n",
    "#@Version: v2.10\n",
    "#@Description: change the anchorcluster color back to mupltiple one\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/9/10 下午1:05:04\n",
    "#@Version: \n",
    "#@Description: change \"fragment_mid\" to \"pos\"\n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/9/4 下午8:51:14\n",
    "#@Version: v2.7\n",
    "#@Description: anchorcluster two color \n",
    "\n",
    "#@Author: Zhuowen Li\n",
    "#@LastEdit: 2021/8/31 下午11:04:38\n",
    "#@Version: v2.6\n",
    "#@Description: \n",
    "#1.mark file columns: chrom start end name color\n",
    "#2.provide optional sorting for the heatmap: maxily keep all reads in the plot\n",
    "\n",
    "\n",
    "\n",
    "#@Author: Zhuowen Licd\n",
    "#@LastEdit: 2021/8/21 下午8:22:05\n",
    "#@Version: v2.4\n",
    "#@Description: \n",
    "#1.step sample according to the combinations of anchors, for cluster heatmap drawing\n",
    "\n",
    "\n",
    "'''\n",
    "Version description_v9: \n",
    "1.\n",
    "output the read matrix after sorted by clustr heatmap\n",
    "\n",
    "Basic Function:\n",
    "-plot porec/hic interactions along with reads.\n",
    "-anchor one site or more to find the site interact with it.(when anchor serval sites, can choose anchormode = and/or)\n",
    "-compare different porec/hic interaction.\n",
    "-normalize with bin infomation.\n",
    "-Modification in bedgraph format(eg. H3K4me1), or special site distribution(eg.DpnII cutting site) listed in format ['chrom'\\t'site'] can be draw along side with same binsize.\n",
    "\n",
    "Attention:\n",
    "Interaction plots (including line or bar plot in porecplot, detail and clusterheat) only count 0/1 (exist or not) for per read in per bin to mask the local interaction.\n",
    "The heatmap would keep the interaction pattern within bins with different degree of color\n",
    "\n",
    "'''\n",
    "\n",
    "import argparse\n",
    "#porec basic explore\n",
    "parser = argparse.ArgumentParser(description='Exploring the interaction of special parts using poreC data. Using the anchor FILE as anchor filtering condition. The pipeline will choose the reads that contact all the anchors in the anchor FILE, which can tell you the level of direct interaction between those anchors.')\n",
    "\n",
    "#porecplot basic infomation\n",
    "parser.add_argument('--align_merge_part',type=str,help='align_merge_part input')\n",
    "parser.add_argument('--keepall',type=str,help='for normalization')\n",
    "parser.add_argument('--region', type=str, help='Input the explore region in bed format, should be written in order. Necessary for porecplot, detail and clusterheat option. The chr columns should be the same format as align file.' )\n",
    "parser.add_argument('--anchor',type=str,help='The anchors you anchor to filter the reads. Necessary for poreplot and detail option')\n",
    "parser.add_argument('--anchormode',type=str,help='anchor reads that catch sites with all anchor points[anchormode = and] or one of then [anchormode = or]')\n",
    "\n",
    "#plots options\n",
    "parser.add_argument('--draw',action = 'store_true',help='Draw or not')\n",
    "parser.add_argument('--mark',type=str,help='Show mark position with color')\n",
    "parser.add_argument('--plotkind',type=str,help='Draw line plot or bar plot of porecplot')\n",
    "parser.add_argument('--genemodel',type=str,nargs='?', const='',help='Draw gene model in porecplot')\n",
    "parser.add_argument('--clusterheat',action='store_true',help='Draw clusterheat or not')\n",
    "parser.add_argument('--anchorcluster',action='store_true',help='Draw anchorcluster or not')\n",
    "\n",
    "#calculation parameter\n",
    "parser.add_argument('--anchor_stat',action = 'store_true',help = 'do anchor statistics')\n",
    "parser.add_argument('--step_sample',type = float,help = 'step sample according to the combinations of anchors, for cluster heatmap drawing')\n",
    "parser.add_argument('--calculate',type=str,help='Choose different calculation methods for interaction value, raw/logy/binnor/binnorlogy')\n",
    "parser.add_argument('--binsize',type=int,help='Decide plotting bin size')\n",
    "parser.add_argument('--ylim',help='set the ylim for distribution plot in porecplot')\n",
    "\n",
    "#other bg_ylimss\n",
    "parser.add_argument('--bedgraphs',nargs='*',help='bedgraph draw, seperated by space in order')\n",
    "parser.add_argument('--bg_labels',nargs='*',help='bedgraph labels, seperated by space in order ')\n",
    "parser.add_argument('--bg_ylims',nargs='*',help='bedgraph ylims, seperated by space in order')\n",
    "\n",
    "#special site distribution\n",
    "parser.add_argument('--sites',nargs='*',help='site draw, seperated by space in order')\n",
    "parser.add_argument('--site_labels',nargs='*',help='site labels, seperated by space in order ')\n",
    "parser.add_argument('--site_ylims',nargs='*',help='site ylims')\n",
    "\n",
    "#output parameter\n",
    "parser.add_argument('--prefix',type=str,help='output file prefix')\n",
    "parser.add_argument('--format',type=str,help='porecplot format')\n",
    "                   \n",
    "args = parser.parse_args()\n",
    "\n",
    "from porecplot_draw_func import *\n",
    "import numpy as np\n",
    "import math\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "font_dirs = [\"/public/home/lizw/software/font\"]\n",
    "font_files = font_manager.findSystemFonts(fontpaths=font_dirs)\n",
    "for font_file in font_files:\n",
    "    font_manager.fontManager.addfont(font_file)\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "plt.rcParams[\"font.size\"] = 20\n",
    "\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap, Normalize\n",
    "\n",
    "\n",
    "\n",
    "region_df = region_reads(args.region)\n",
    "anchor = pd.read_csv(args.anchor,sep=\"\\t\",header=None,usecols=[0,1,2,3],names=['chrom','start','end','name'],converters={'chrom':str, 'start':int, 'end':int,'name':str},comment=\"#\")\n",
    "df_merge_part = pd.read_parquet(args.align_merge_part)\n",
    "anchor_reads_detail_df = anchor_detail(anchor,df_merge_part,args.prefix)\n",
    "\n",
    "if args.anchor_stat == True:\n",
    "    anchor_statics(anchor_reads_detail_df,args.prefix)\n",
    "\n",
    "merge_keepall_part = pd.read_parquet(args.keepall)\n",
    "merge_keepall_part.rename(columns={'pos':'site'},inplace=True)\n",
    "\n",
    "anchor_interval_count,anchor_filter_merge,anchor_and_list = anchor_reads(anchor,anchor_reads_detail_df,df_merge_part,region_df,'pos',args.binsize,args.anchormode,args.prefix,True,step_sample=args.step_sample)\n",
    "anchor_binnor_merge_all_sort = anchor_calcu(merge_keepall_part,anchor_interval_count,region_df,args.binsize,args.prefix,True)\n",
    "\n",
    "\n",
    "if args.anchorcluster:\n",
    "    viridis = cm.get_cmap('OrRd', 6)\n",
    "    newcolors = viridis(np.linspace(0, 1, 256))\n",
    "    white = np.array([255/255, 255/255, 255/255, 1])\n",
    "    newcolors[:50, :] = white\n",
    "    newcmp1 = ListedColormap(newcolors)\n",
    "\n",
    "    anchor_and_detail = anchor_reads_detail_df.query('read_name in @anchor_and_list')\n",
    "    read_anchor_matrix = anchor_and_detail.groupby(['read_name','anchor_name'])['pos'].count().unstack().fillna(0).reindex(anchor['name'].to_list(), axis=1)\n",
    "    #make a only 0_1 matrix for sort, to make the same combinations together\n",
    "    read_anchor_matrix_01 = read_anchor_matrix != 0\n",
    "    read_anchor_matrix_01_sort = read_anchor_matrix_01.sort_values(by=anchor['name'].to_list())\n",
    "    #make the read list of comb sort as ref,and sort the orginal matrix(contain the information of locus self interaction within one read)\n",
    "    read_comb_sort_list = read_anchor_matrix_01_sort.reset_index().iloc[:,0].to_list()\n",
    "    read_anchor_matrix_sort=read_anchor_matrix.reindex(read_comb_sort_list,axis=0)\n",
    "    fig_anclu, ax_anclu = plt.subplots(1,figsize=(20,20))\n",
    "    g_anclu = sns.heatmap(read_anchor_matrix_sort.reset_index().iloc[:,1:],vmin=0-0.5,vmax=5+0.5,ax=ax_anclu,cmap=newcmp1,cbar=False)\n",
    "    #g_anclu.ax_heatmap.tick_params(tick2On=False, labelsize=False)\n",
    "#     ax_anclu.set_xlabels('')\n",
    "#     ax_anclu.set_yticklabels([])\n",
    "    plt.savefig(args.prefix+'_anchorcluster.'+ args.format,format = args.format,dpi=300,bbox_inches = 'tight')\n",
    "\n",
    "if args.draw == True:\n",
    "    #define the absolute length of each part, and calculate the proportion, so the main plot would stay the same while other tracks added\n",
    "    heat_concat,read_concat = heat_df(anchor_binnor_merge_all_sort,anchor_filter_merge)\n",
    "    heat_concat.columns = list(range(0,len(heat_concat.columns)))\n",
    "    bg_len = len(args.bedgraphs) if args.bedgraphs else 0\n",
    "    site_len = len(args.sites) if args.sites else 0 \n",
    "    yflen = 12+bg_len+site_len\n",
    "    heat_concat_sort = heat_concat.sort_values(by=heat_concat.columns.to_list())\n",
    "    viridis = cm.get_cmap('BuPu', 6)\n",
    "    newcolors2 = viridis(np.linspace(0, 1, 256))\n",
    "    white = np.array([255/255, 255/255, 255/255, 1])\n",
    "    newcolors2[:50, :] = white\n",
    "    newcmp2 = ListedColormap(newcolors2)\n",
    "\n",
    "    if len(heat_concat_sort) >= 10:\n",
    "        cluster_top_para = 4.2\n",
    "        cluster_bottom_para = colorbar_bottom_para = 10.5\n",
    "        colorbar_top_para = 5\n",
    "        clustermap = sns.clustermap(heat_concat_sort,figsize=(30,yflen),cmap = newcmp2,col_cluster=False,row_cluster=args.clusterheat,vmin=0-0.5,vmax=5+0.5,dendrogram_ratio=0.1,cbar_pos=None)\n",
    "        clustermap.ax_heatmap.tick_params(tick2On=False, labelsize=False)\n",
    "        cluster_order = clustermap.dendrogram_row.reordered_ind\n",
    "        #output the clustermatrix\n",
    "        heat_concat_sort_cluster = heat_concat_sort.reindex(cluster_order)\n",
    "        heat_concat_sort_cluster.to_csv(args.prefix+'_heat_matrix.csv')\n",
    "        \n",
    "        ax_cluster = clustermap.fig.gca()\n",
    "        ax_cluster.set_xticks([])\n",
    "        #clustermap.gs.update(top=1-cluster_top_para/yflen, bottom=1-cluster_bottom_para/yflen,left=-0.00000000000000000001,right = 0.9)\n",
    "        clustermap.gs.update(top=1-cluster_top_para/yflen, bottom=1-cluster_bottom_para/yflen,left=0.0077,right = 0.9)\n",
    "\n",
    "        #mainplot\n",
    "        mainplot = clustermap.fig\n",
    "        #colorbar\n",
    "        gs1 = GridSpec(1,1,top=1-colorbar_top_para/yflen, bottom=1-colorbar_bottom_para/yflen,left = 0.95, right = 1) \n",
    "        ax_color = mainplot.add_subplot(gs1[0])\n",
    "        ax_color.set_yticklabels([])\n",
    "        remove_draw(ax_color,tick=True,spine=True,locator=True)\n",
    "        cbnorm = Normalize(vmin=0-0.5,vmax=5+0.5,clip=False)\n",
    "        cb = plt.colorbar(cm.ScalarMappable(norm=cbnorm, cmap=newcmp2),fraction=1,ax=ax_color,ticks=np.arange(6))\n",
    "\n",
    "    else:\n",
    "        cluster_top_para = cluster_bottom_para = colorbar_top_para = colorbar_bottom_para = 0\n",
    "        with open (f'{args.prefix}.report','a+') as report:\n",
    "            report.write('Reads count lower than 10! Remove cluster heatmap.')\n",
    "        mainplot = plt.figure()\n",
    "\n",
    "\n",
    "    gs2 = GridSpec(1,1, top=1,bottom=1-4.8/yflen,left = 0.10,right = 0.9)\n",
    "    #create axes within this new gridspec\n",
    "    ax_binsum = mainplot.add_subplot(gs2[0])\n",
    "    porec_draw(anchor_binnor_merge_all_sort,ax_binsum,args.calculate,args.plotkind,args.ylim)\n",
    "\n",
    "    below_cluster_para = cluster_bottom_para + 0.5\n",
    "    #bedgraph information\n",
    "    if args.bedgraphs:\n",
    "        bg_top = 1-below_cluster_para/yflen\n",
    "        bg_bottom = 1-(bg_len+below_cluster_para)/yflen\n",
    "        gs3 = GridSpec(bg_len,1,top=bg_top,bottom=bg_bottom,left=0.10,right = 0.9,hspace=0.2)\n",
    "        track_draw(args.bedgraphs,args.bg_labels,args.bg_ylims,mainplot,region_df,args.binsize,gs3,0,'bg') \n",
    "\n",
    "\n",
    "    if args.sites:\n",
    "        site_top = 1-(bg_len+0.2+below_cluster_para)/yflen\n",
    "        site_bottom = 1-(bg_len+0.2+site_len+below_cluster_para)/yflen\n",
    "        gs4 = GridSpec(site_len,1,top=site_top,bottom=site_bottom,left=0.10,right = 0.9,hspace=0.2)\n",
    "        track_draw(args.sites,args.site_labels,args.site_ylims,mainplot,region_df,args.binsize,gs4,bg_len,'site')   \n",
    "\n",
    "    #anchor and mark information\n",
    "    region_df_cum = region_df.copy()\n",
    "    region_df_cum['cum'] = anchor_binnor_merge_all_sort.groupby('region_index').count()['index'].cumsum()\n",
    "    region_figlensum = anchor_binnor_merge_all_sort.groupby('region_index').count()['index'].sum()\n",
    "    gs5 = GridSpec(1,region_figlensum,top=1,bottom=0.0,left=0.10,right = 0.9)\n",
    "    \n",
    "    kw = dict()\n",
    "    if args.genemodel:\n",
    "        gene_isoform = pd.read_csv(args.genemodel, sep='\\t', names=['chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts'],comment=\"#\")\n",
    "        kw['gene'] = gene_isoform\n",
    "    if args.mark:\n",
    "        mark = pd.read_csv(args.mark,sep=\"\\t\",header=None,names=['chrom','start','end','name','color'],converters={'chrom':str, 'start':int, 'end':int,'color':str},comment=\"#\")\n",
    "        kw['mark'] = mark\n",
    "\n",
    "    rec_draw(region_df_cum,mainplot,anchor,args.binsize,gs5,yflen,args.anchormode,**kw)\n",
    "    plt.savefig(args.prefix+'.'+ args.format,format = args.format,dpi=500,bbox_inches = 'tight')\n",
    "\n",
    "with open (f'{args.prefix}.report','a+') as report:\n",
    "    report.write('Succesfully finish!')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-02T03:21:41.313262Z",
     "start_time": "2021-09-02T03:21:36.522178Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 2, 3, 4]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(0,5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-22T08:05:49.822087Z",
     "start_time": "2021-09-22T08:05:49.274665Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-22T08:12:21.068751Z",
     "start_time": "2021-09-22T08:10:43.333361Z"
    }
   },
   "outputs": [],
   "source": [
    "df_merge_part = pd.read_parquet('/public/home/lizw/task/pore_c/porec_1000_filter_mainchr_result/porecplot_v3/porec_pre_hic_v2.8/DpnII_hic.hic_merge_part.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-22T08:12:24.094983Z",
     "start_time": "2021-09-22T08:12:24.049493Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>read_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44279054</td>\n",
       "      <td>SRR1029605.1.10000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1897711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>44279054</td>\n",
       "      <td>SRR1029605.1.10000000</td>\n",
       "      <td>3</td>\n",
       "      <td>1905603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>46458481</td>\n",
       "      <td>SRR1029605.1.100000002</td>\n",
       "      <td>3</td>\n",
       "      <td>6817354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>46458481</td>\n",
       "      <td>SRR1029605.1.100000002</td>\n",
       "      <td>3</td>\n",
       "      <td>4556657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46557315</td>\n",
       "      <td>SRR1029605.1.100000005</td>\n",
       "      <td>3</td>\n",
       "      <td>9571905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185405831</th>\n",
       "      <td>37650836</td>\n",
       "      <td>SRR1029605.1.99999983</td>\n",
       "      <td>2</td>\n",
       "      <td>17851275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185405832</th>\n",
       "      <td>11377168</td>\n",
       "      <td>SRR1029605.1.99999992</td>\n",
       "      <td>1</td>\n",
       "      <td>17200095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185405833</th>\n",
       "      <td>11377168</td>\n",
       "      <td>SRR1029605.1.99999992</td>\n",
       "      <td>1</td>\n",
       "      <td>17200474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185405834</th>\n",
       "      <td>1971766</td>\n",
       "      <td>SRR1029605.1.99999994</td>\n",
       "      <td>1</td>\n",
       "      <td>2207396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185405835</th>\n",
       "      <td>1971766</td>\n",
       "      <td>SRR1029605.1.99999994</td>\n",
       "      <td>1</td>\n",
       "      <td>4469018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>185405836 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              index               read_name  chrom       pos\n",
       "0          44279054   SRR1029605.1.10000000      3   1897711\n",
       "1          44279054   SRR1029605.1.10000000      3   1905603\n",
       "2          46458481  SRR1029605.1.100000002      3   6817354\n",
       "3          46458481  SRR1029605.1.100000002      3   4556657\n",
       "4          46557315  SRR1029605.1.100000005      3   9571905\n",
       "...             ...                     ...    ...       ...\n",
       "185405831  37650836   SRR1029605.1.99999983      2  17851275\n",
       "185405832  11377168   SRR1029605.1.99999992      1  17200095\n",
       "185405833  11377168   SRR1029605.1.99999992      1  17200474\n",
       "185405834   1971766   SRR1029605.1.99999994      1   2207396\n",
       "185405835   1971766   SRR1029605.1.99999994      1   4469018\n",
       "\n",
       "[185405836 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge_part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-27T06:23:56.467536Z",
     "start_time": "2021-09-27T06:23:55.940354Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-27T06:24:27.662028Z",
     "start_time": "2021-09-27T06:24:21.600165Z"
    }
   },
   "outputs": [],
   "source": [
    "df = pd.read_parquet('/public/home/lizw/task/pore_c/porec_1000_filter_mainchr_result/porecplot_v3/porec_pre_v2.9/DpnII_run08.order_2_merge_part.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-09-27T06:24:29.384737Z",
     "start_time": "2021-09-27T06:24:29.310201Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chrom</th>\n",
       "      <th>read_name</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2910745e-eb49-4901-a1f2-4491ede1d0d7</td>\n",
       "      <td>20076841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2910745e-eb49-4901-a1f2-4491ede1d0d7</td>\n",
       "      <td>18952877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2910745e-eb49-4901-a1f2-4491ede1d0d7</td>\n",
       "      <td>19596981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2910745e-eb49-4901-a1f2-4491ede1d0d7</td>\n",
       "      <td>17085746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>2910745e-eb49-4901-a1f2-4491ede1d0d7</td>\n",
       "      <td>18879034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22810633</th>\n",
       "      <td>36378731</td>\n",
       "      <td>2</td>\n",
       "      <td>05d0ead4-62e1-40fe-94f7-4a3b7165598d</td>\n",
       "      <td>972242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22810634</th>\n",
       "      <td>36378732</td>\n",
       "      <td>2</td>\n",
       "      <td>05d0ead4-62e1-40fe-94f7-4a3b7165598d</td>\n",
       "      <td>4964832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22810635</th>\n",
       "      <td>36378734</td>\n",
       "      <td>2</td>\n",
       "      <td>05d0ead4-62e1-40fe-94f7-4a3b7165598d</td>\n",
       "      <td>4703718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22810636</th>\n",
       "      <td>36378737</td>\n",
       "      <td>2</td>\n",
       "      <td>4307c964-4e17-4ed0-9e4d-aca5ea93f43c</td>\n",
       "      <td>19407635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22810637</th>\n",
       "      <td>36378738</td>\n",
       "      <td>5</td>\n",
       "      <td>4307c964-4e17-4ed0-9e4d-aca5ea93f43c</td>\n",
       "      <td>6633027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22810638 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             index chrom                             read_name       pos\n",
       "0                1     3  2910745e-eb49-4901-a1f2-4491ede1d0d7  20076841\n",
       "1                2     1  2910745e-eb49-4901-a1f2-4491ede1d0d7  18952877\n",
       "2                3     3  2910745e-eb49-4901-a1f2-4491ede1d0d7  19596981\n",
       "3                5     2  2910745e-eb49-4901-a1f2-4491ede1d0d7  17085746\n",
       "4                6     3  2910745e-eb49-4901-a1f2-4491ede1d0d7  18879034\n",
       "...            ...   ...                                   ...       ...\n",
       "22810633  36378731     2  05d0ead4-62e1-40fe-94f7-4a3b7165598d    972242\n",
       "22810634  36378732     2  05d0ead4-62e1-40fe-94f7-4a3b7165598d   4964832\n",
       "22810635  36378734     2  05d0ead4-62e1-40fe-94f7-4a3b7165598d   4703718\n",
       "22810636  36378737     2  4307c964-4e17-4ed0-9e4d-aca5ea93f43c  19407635\n",
       "22810637  36378738     5  4307c964-4e17-4ed0-9e4d-aca5ea93f43c   6633027\n",
       "\n",
       "[22810638 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
